core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -318,6 +323,16 @@ private Object processWithExceptionHandling(Expression expression, Object contex
             }
         }
 
+        private boolean canUseCache(Object context)
+        {
+            // We are using WeakHashMap as cache, that's why we can't depend on an intermediate object
+            // that consists of expression as well as context as a key. This is because intermediate object could
+            // get GCed even when expression still exists which will cause cache miss. So, to solve this problem we
+            // can only cache when context is noop. It'll also give us pretty good perf benefits because in most of
+            // places optimize is called with noop symbol resolver.
```

Comment:
Can same expression (by identity) be evaluated to something else with different `SymbolResolver`? That would seem unlikely, no?

Reply:
> Can same expression (by identity) be evaluated to something else with different SymbolResolver?

Yes, it is possible. There are some tests which were failing before that's why I did this change. For ex: `io.trino.sql.planner.AbstractPredicatePushdownTest#testNormalizeOuterJoinToInner`

So. specifically for this test, the `SymbolReference` is called twice for optimization one without context and another one with the context where the symbol tends to be null.

Another thing that I want to point out is that `WeakHashmap` isn't based on identity, although GC happens on Identity but the get by key works on equals and hashing.

Reply:
> Another thing that I want to point out is that `WeakHashmap` isn't based on identity, although GC happens on Identity but the get by key works on equals and hashing.

Good point. I was blindsided by `com.google.common.cache.CacheBuilder#weakKeys` implying the keys are compared with `==`. Indeed, `WeakHashMap` is not applicable here because `Expression.equals` is not a good semantics for the expression cache.

Reply:
In that case, I'm thinking to implement it using `IdentityHashMap` because anyways `ExpressionInterpreter` is created per query planning. So, it will get GCed, hence there shouldn't be any memory leak right?

Reply:
planning can create new Expressions and old, unused expressions should be eligible for GC

Reply:
So there are two option IMO
1. Create a new map -> WeakIdentityHashMap which doesn't exist in JDK.
2. Use `com.google.common.cache.CacheBuilder#weakKeys` but it has extra cost due to being concurrent.

Reply:
> 2\. Use `com.google.common.cache.CacheBuilder#weakKeys` but it has extra cost due to being concurrent.

or `com.google.common.collect.MapMaker#weakKeys`
(also concurrent map)

> 1\. Create a new map -> WeakIdentityHashMap which doesn't exist in JDK.

Sounds complicated.

Reply:
@gaurav8297 and I discussed this offline. The conclusion was that we **may** be able to use equality-based semantics for Expressions. Identity-based comparisons are a **must** for analyzer and planner, but they **may** be unnecessary during query optimization.

@martint please chime in here.

Reply:
@findepi I ran benchmarks with this `com.google.common.collect.MapMaker#weakKeys` again. The overhead is pretty negligible. So maybe it's better to simply use this?